### PR TITLE
OCI-based devifle registry support

### PR DESCRIPTION
**Pull request by GeekArthur**
Signed-off-by: jingfu wang <jingfwan@redhat.com>

**What type of PR is this?**
/kind feature

**What does this PR do / why we need it**:
This PR implements the OCI-based devfile registry support, more specifically:
1. Consume devfile registry library or REST API (list devfile components + create devfile component)
2. Change the default devfile registry URL to OCI-based public devfile registry

**Which issue(s) this PR fixes**:

Related issues:
1. https://github.com/devfile/api/issues/367
2. https://github.com/openshift/odo/issues/4504

**PR acceptance criteria**:

- [x] Unit test 

- [x] Integration test 

- [x] Documentation 

- [x] I have read the [test guidelines](https://github.com/openshift/odo/blob/master/docs/dev/test-architecture.adoc)

**How to test changes / Special notes to the reviewer**:
1. Run `odo registry add OCI https://registry.devfile.io` to add OCI-based public devfile registry if it's not in the registry list
2. Run `odo catalog list components` then verify the devfiles from OCI-based public devfile registry can be listed
3. Run `odo create nodejs --registry OCI` then verify the nodejs devfile can be downloaded from OCI-based public devfile registry

### CODE REVIEW

File: pkg/odo/cli/component/create.go
Comment (johnmcollier) [inline pkg/odo/cli/component/create.go:815]: This could cause problems for stacks with multiple artifacts in the registry, as this will only download the devfile, rather than everything.

Comment (GeekArthur) [reply pkg/odo/cli/component/create.go:815]: I think this has been discussed before, for the time being odo wouldn't support download the whole stack. It's consistent with today's Github-based registry, for example the nodejs stack from https://github.com/odo-devfiles/registry/tree/master/devfiles/nodejs has multiple artifacts but if we run `odo create nodejs` it only downloads the devfile.

Comment (johnmcollier) [reply pkg/odo/cli/component/create.go:815]: Yes, but part of the reason we're moving to the new registry format is so that those stacks with multiple files can be supported. Shouldn't it be pretty straightforward to change the existing logic just to call `registryLibrary.PullStackFromRegistry(registry, stack)` instead?

CC @kadel

Comment (GeekArthur) [reply pkg/odo/cli/component/create.go:815]: Actually it's more complicated than just switch to call another function, currently for OCI-based registry support we use registry REST API instead of registry library (eg. `registryLibrary.PullStackFromRegistry(registry, stack)`), this can align with existing Github-based registry and use the existing odo devfile parser/API (eg. `DownloadFileInMemoryWithCache` which is the wrapper of `HTTPGetRequest`) for component creation. If we want to switch to use registry library we need to do the following work/design:
1. Add additional option to `odo create` for downloading the whole stack
2. Modify the registry library and current odo functions to ensure all component creation features won't be affected, this should at least include:
   2.1. Allow registry library support customized context/root folder
   2.2. Ensure download starter project works well
   2.3. Implement the cache mechanism as we can't reuse the existing odo function `DownloadFileInMemoryWithCache`

But I think the key question is do we want to let odo support downloading the whole stack for now, IMO the Github-based registry and OCI-based registry should be consistent.

Comment (kadel) [reply pkg/odo/cli/component/create.go:815]: > I think this has been discussed before, for the time being odo wouldn't support download the whole stack. It's consistent with today's Github-based registry, for example the nodejs stack from https://github.com/odo-devfiles/registry/tree/master/devfiles/nodejs has multiple artifacts but if we run `odo create nodejs` it only downloads the devfile.

The extra files in that nodejs stack are for building the image but they are not required for devfile.yaml to function properly. 
This is not a good example and I don't think that we should store a source like this with devfile in the registry. 

If the stack has multiple files, and those files are required for devfile.yaml to function properly odo will need to download everything.

Comment (GeekArthur) [reply pkg/odo/cli/component/create.go:815]: > The extra files in that nodejs stack are for building the image but they are not required for devfile.yaml to function properly.
This is not a good example and I don't think that we should store a source like this with devfile in the registry.

As far as I know, currently all extra files are about outer loop configuration (eg. Dockerfile, k8s manifest), I think the nodejs is the only one has extra files within the stack in our registry for now.

> If the stack has multiple files, and those files are required for devfile.yaml to function properly odo will need to download everything. If we still net

Yes, if files are required for devfile.yaml, currently odo can download all of them from the OCI-based registry with this PR, which is similar to Github-based registry.

Given that, I think for the time being I will make the Github-based registry and OCI-based registry consistent, they can both download the files that required devfile.yaml, and we can extend the functionality once we decide how to host the outer loop files.

**Comment by kadel** [comment]
We need to find a way how to change `DefaultDevfileRegistry` for the current users.
Existing odo users will already have the existing `DefaultDevfileRegistry` record in `preference.yaml`, odo needs to update it to the new registry, otherwise all current users will get stuck with default GitHub based registry.

**Comment by johnmcollier** [comment]
> We need to find a way how to change DefaultDevfileRegistry for the current users.
Existing odo users will already have the existing DefaultDevfileRegistry record in preference.yaml, odo needs to update it to the new registry, otherwise, all current users will get stuck with the default GitHub based registry.

Yup, agreed. Some useful discussion in https://github.com/openshift/odo/issues/3298 about this as well that might provide some insight on how we can do this

**Comment by johnmcollier** [comment]
@GeekArthur The `odo catalog describe` tests are failing due to this error:
```
 [odo] Starter Projects:
[odo] I0330 15:04:12.105604   59639 validate.go:50] Successfully validated devfile sections
[odo] ---
[odo] name: nodejs-starter
[odo] attributes: {}
[odo] description: ""
[odo] subdir: ""
[odo] projectsource:
[odo]   sourcetype: ""
[odo]   git:
[odo]     gitlikeprojectsource:
[odo]       commonprojectsource: {}
[odo]       checkoutfrom: null
[odo]       remotes:
[odo]         origin: https://github.com/odo-devfiles/nodejs-ex.git
[odo]   github: null
[odo]   zip: null
[odo]   custom: null
[odo] 
[odo] * Registry: DefaultDevfileRegistry
[odo]  ✗  Get https://registry.devfile.iodevfile-catalog/nodejs:latest: dial tcp: lookup registry.devfile.iodevfile-catalog on 172.30.0.10:53: no such host 
```

You need to update the `GetDevfile` function to properly handle cases where the devfile is in an OCI registry:

https://github.com/openshift/odo/blob/master/pkg/odo/cli/catalog/describe/component.go#L197-L209

**Comment by GeekArthur** [comment]
> > We need to find a way how to change DefaultDevfileRegistry for the current users.
> > Existing odo users will already have the existing DefaultDevfileRegistry record in preference.yaml, odo needs to update it to the new registry, otherwise, all current users will get stuck with the default GitHub based registry.
> 
> Yup, agreed. Some useful discussion in #3298 about this as well that might provide some insight on how we can do this

Yeah agree, I also wanna link Maysun's issue here, thanks John.

File: pkg/odo/cli/component/create.go
Comment (GeekArthur) [reply pkg/odo/cli/component/create.go]: Based on the cabal call discussion, we will make odo download the whole stack by default when user run `odo create` to create devfile component.

**Comment by GeekArthur** [comment]
> @GeekArthur The `odo catalog describe` tests are failing due to this error:
> 
> ```
>  [odo] Starter Projects:
> [odo] I0330 15:04:12.105604   59639 validate.go:50] Successfully validated devfile sections
> [odo] ---
> [odo] name: nodejs-starter
> [odo] attributes: {}
> [odo] description: ""
> [odo] subdir: ""
> [odo] projectsource:
> [odo]   sourcetype: ""
> [odo]   git:
> [odo]     gitlikeprojectsource:
> [odo]       commonprojectsource: {}
> [odo]       checkoutfrom: null
> [odo]       remotes:
> [odo]         origin: https://github.com/odo-devfiles/nodejs-ex.git
> [odo]   github: null
> [odo]   zip: null
> [odo]   custom: null
> [odo] 
> [odo] * Registry: DefaultDevfileRegistry
> [odo]  ✗  Get https://registry.devfile.iodevfile-catalog/nodejs:latest: dial tcp: lookup registry.devfile.iodevfile-catalog on 172.30.0.10:53: no such host 
> ```
> 
> You need to update the `GetDevfile` function to properly handle cases where the devfile is in an OCI registry (specifically the call to `devfile.ParseFromURLAndValidate`):
> 
> https://github.com/openshift/odo/blob/12b03e5924b1db8ea1634ecf22e9aab215bc7ab5/pkg/odo/cli/catalog/describe/component.go#L197-L209

Yep, will fix that with other test automation failures.

**Comment by elsony** [comment]
@GeekArthur , @kadel and I have discussed the default registry migration. We'll just take a look at the existing preference, if there is a default entry that points to the old repo, we'll just migrate the registry URL to the new URL silently in the preference file without prompting the user.

**Comment by johnmcollier** [comment]
Looking like CI infra issues? FWIW, I hit similar issues earlier today with provisioning a test cluster using the OpenShift CI with cluster-bot, so those could be related:
```
 level=error msg=Error: Error creating IAM Role ci-op-nrb66cyk-ba78b-85jvp-bootstrap-role: LimitExceeded: Cannot exceed quota for RolesPerAccount: 1000
level=error msg=	status code: 409, request id: 5bd8b476-f78f-4c89-a4bd-a5b0966b26d0
level=error
level=error msg=  on ../tmp/openshift-install-295422802/bootstrap/main.tf line 51, in resource "aws_iam_role" "bootstrap":
level=error msg=  51: resource "aws_iam_role" "bootstrap" {
level=error
level=error
level=error
level=error msg=Error: Error creating IAM Role ci-op-nrb66cyk-ba78b-85jvp-worker-role: LimitExceeded: Cannot exceed quota for RolesPerAccount: 1000
level=error msg=	status code: 409, request id: b0a50cc7-4ff2-4351-a030-7b11ce76d806
level=error
level=error msg=  on ../tmp/openshift-install-295422802/iam/main.tf line 13, in resource "aws_iam_role" "worker_role":
level=error msg=  13: resource "aws_iam_role" "worker_role" {
level=error
level=error
level=error
level=error msg=Error: Error creating IAM Role ci-op-nrb66cyk-ba78b-85jvp-master-role: LimitExceeded: Cannot exceed quota for RolesPerAccount: 1000
level=error msg=	status code: 409, request id: 502351a7-55d7-4821-b8ad-400cec346294
level=error
level=error msg=  on ../tmp/openshift-install-295422802/master/main.tf line 17, in resource "aws_iam_role" "master_role":
level=error msg=  17: resource "aws_iam_role" "master_role" {
level=error
level=error
level=error
level=error msg=Error: Error creating VPC: VpcLimitExceeded: The maximum number of VPCs has been reached.
level=error msg=	status code: 400, request id: 33a54f1e-c7a0-4b13-9639-25d3ecf59d49
level=error
level=error msg=  on ../tmp/openshift-install-295422802/vpc/vpc.tf line 6, in resource "aws_vpc" "new_vpc":
level=error msg=   6: resource "aws_vpc" "new_vpc" {
level=error
level=error
level=fatal msg=failed to fetch Cluster: failed to generate asset "Cluster": failed to create cluster: failed to apply Terraform: failed to complete the change 
```

**Review by johnmcollier — CHANGES_REQUESTED** [review]
New changes generally look good with using the registry library to pull the stacks now. One comment on something that I think we need to fix on probably the registry library side.

When I do `odo create nodejs ...` with the OCI registry, we get a lot of output from oras:
```
 ✓  Validating if devfile name is correct [149329ns]
2021/04/06 13:07:22 Pulling stack nodejs from devfile-registry-johnmcollier-code.apps.sandbox.x8i5.p1.openshiftapps.com/devfile-catalog/nodejs:latest with allowed media types [application/vnd.devfileio.devfile.layer.v1 image/png image/svg+xml application/vnd.devfileio.vsx.layer.v1.tar application/x-tar]...
WARN[0000] reference for unknown type: application/vnd.devfileio.devfile.layer.v1  digest="sha256:b72d04e51f3df4a1411b9d387a8cd80e4d9e5f2f410b7c8849c9e4a1e5ec148f" mediatype=application/vnd.devfileio.devfile.layer.v1 size=1259
WARN[0000] unknown type: application/vnd.devfileio.devfile.config.v2+json 
WARN[0000] reference for unknown type: application/x-tar  digest="sha256:0f972e76bc01251e06e7ea4d2a3d606e58f4334a4b9366409830cc1cb82c3b51" mediatype=application/x-tar size=864
2021/04/06 13:07:22 Pulled stack nodejs from devfile-registry-johnmcollier-code.apps.sandbox.x8i5.p1.openshiftapps.com/devfile-catalog/nodejs:latest with digest sha256:5fd4dcfe56f451412136df968e08237d77f88a5770c86c0a2eed7caf60be3355
```

It looks like if the registry library uses oras' context instead of the generic go context in calls to oras (via importing `orasctx "github.com/deislabs/oras/pkg/context"`), we can avoid that messy output.

**Comment by GeekArthur** [comment]
> New changes generally look good with using the registry library to pull the stacks now. One comment on something that I think we need to fix on probably the registry library side.
> 
> When I do `odo create nodejs ...` with the OCI registry, we get a lot of output from oras:
> 
> ```
>  ✓  Validating if devfile name is correct [149329ns]
> 2021/04/06 13:07:22 Pulling stack nodejs from devfile-registry-johnmcollier-code.apps.sandbox.x8i5.p1.openshiftapps.com/devfile-catalog/nodejs:latest with allowed media types [application/vnd.devfileio.devfile.layer.v1 image/png image/svg+xml application/vnd.devfileio.vsx.layer.v1.tar application/x-tar]...
> WARN[0000] reference for unknown type: application/vnd.devfileio.devfile.layer.v1  digest="sha256:b72d04e51f3df4a1411b9d387a8cd80e4d9e5f2f410b7c8849c9e4a1e5ec148f" mediatype=application/vnd.devfileio.devfile.layer.v1 size=1259
> WARN[0000] unknown type: application/vnd.devfileio.devfile.config.v2+json 
> WARN[0000] reference for unknown type: application/x-tar  digest="sha256:0f972e76bc01251e06e7ea4d2a3d606e58f4334a4b9366409830cc1cb82c3b51" mediatype=application/x-tar size=864
> 2021/04/06 13:07:22 Pulled stack nodejs from devfile-registry-johnmcollier-code.apps.sandbox.x8i5.p1.openshiftapps.com/devfile-catalog/nodejs:latest with digest sha256:5fd4dcfe56f451412136df968e08237d77f88a5770c86c0a2eed7caf60be3355
> ```
> 
> It looks like if the registry library uses oras' context instead of the generic go context in calls to oras (via importing `orasctx "github.com/deislabs/oras/pkg/context"`), we can avoid that messy output.

Good suggestion, I can give a try on registry library side.

**Comment by GeekArthur** [comment]
> > New changes generally look good with using the registry library to pull the stacks now. One comment on something that I think we need to fix on probably the registry library side.
> > When I do `odo create nodejs ...` with the OCI registry, we get a lot of output from oras:
> > ```
> >  ✓  Validating if devfile name is correct [149329ns]
> > 2021/04/06 13:07:22 Pulling stack nodejs from devfile-registry-johnmcollier-code.apps.sandbox.x8i5.p1.openshiftapps.com/devfile-catalog/nodejs:latest with allowed media types [application/vnd.devfileio.devfile.layer.v1 image/png image/svg+xml application/vnd.devfileio.vsx.layer.v1.tar application/x-tar]...
> > WARN[0000] reference for unknown type: application/vnd.devfileio.devfile.layer.v1  digest="sha256:b72d04e51f3df4a1411b9d387a8cd80e4d9e5f2f410b7c8849c9e4a1e5ec148f" mediatype=application/vnd.devfileio.devfile.layer.v1 size=1259
> > WARN[0000] unknown type: application/vnd.devfileio.devfile.config.v2+json 
> > WARN[0000] reference for unknown type: application/x-tar  digest="sha256:0f972e76bc01251e06e7ea4d2a3d606e58f4334a4b9366409830cc1cb82c3b51" mediatype=application/x-tar size=864
> > 2021/04/06 13:07:22 Pulled stack nodejs from devfile-registry-johnmcollier-code.apps.sandbox.x8i5.p1.openshiftapps.com/devfile-catalog/nodejs:latest with digest sha256:5fd4dcfe56f451412136df968e08237d77f88a5770c86c0a2eed7caf60be3355
> > ```
> > 
> > 
> > It looks like if the registry library uses oras' context instead of the generic go context in calls to oras (via importing `orasctx "github.com/deislabs/oras/pkg/context"`), we can avoid that messy output.
> 
> Good suggestion, I can give a try on registry library side.

@johnmcollier PR to fix on registry library side: https://github.com/devfile/registry-support/pull/42, please review.

**Comment by valaparthvi** [comment]
@GeekArthur Can you please update this PR branch with the master? A few important changes are missing from this branch and they're most probably sending out wrong telemetry data when you run something locally.

**Comment by GeekArthur** [comment]
> @GeekArthur Can you please update this PR branch with the master? A few important changes are missing from this branch and they're most probably sending out wrong telemetry data when you run something locally.



> @GeekArthur Can you please update this PR branch with the master? A few important changes are missing from this branch and they're most probably sending out wrong telemetry data when you run something locally.

@valaparthvi Sure thing, I am waiting for a dependency PR, once it's merged today I can rebased this PR onto master with some fixes for test cases.

**Comment by kadel** [comment]
I've been playing with this PR today a bit. I haven't found anything else except the issues that were already raised:
- updating existing `DefaultDevfileRegistry` key in users preference file
- warning about the unknown media type

I think that once those get fixed we can merge this. Good work @GeekArthur

**Comment by GeekArthur** [comment]
> I've been playing with this PR today a bit. I haven't found anything else except the issues that were already raised:
> 
>     * updating existing `DefaultDevfileRegistry` key in users preference file
> 
>     * warning about the unknown media type
> 
> 
> I think that once those get fixed we can merge this. Good work @GeekArthur

Thanks @kadel for reviewing the PR.

The second point is resolved in my latest commit. Currently I am working on the first point to provide a auto migration mechanism for OCI-based registry.

**Comment by girishramnani** [comment]
Is this a genuine failure?

**Comment by GeekArthur** [comment]
> Is this a genuine failure?

@girishramnani No, it's test infra issue, I posted it on odo slack channel: https://kubernetes.slack.com/archives/C01D6L2NUAG/p1618243057007700

**Comment by GeekArthur** [comment]
@johnmcollier @kadel All comments addressed, I already discussed with Tomas, due to the test infra issues, the PR can be merged once it satisfies the following requirements:
1. Has `approve` and `lgtm` labels
2. Pass `unit` and `ci/prow/v4.7-integration-e2e` tests

Please review again, thanks.

**Review by johnmcollier — APPROVED** [review]
No longer a reviewer on this repo, but approving from the devfile side of things

/lgtm

Edit: Didn't expect that to do anything 😅

**Comment by kadel** [comment]
> Edit: Didn't expect that to actually add the label 😅 , thought you had to be in the OWNERS file?

wow. Interesting. You still have to write permission on this repo, that might be why. But this is new to me :-)

**Comment by kadel** [comment]
/approve

I have one small issue with the current migration approach. But that can be addressed separately.
Opened https://github.com/openshift/odo/issues/4626  to track it

**Pull request merged**
